File: mifugo-care/app/src/main/java/com/mifugocare/app/database/entities/Disease.java
package com.mifugocare.app.database.entities;

import androidx.room.Entity;
import androidx.room.PrimaryKey;

@Entity(tableName = "diseases")
public class Disease {
    @PrimaryKey(autoGenerate = true)
    public int diseaseId;
    
    public String diseaseName;
    public String scientificName;
    public String species; // Which livestock species this affects
    public String symptoms;
    public String causes;
    public String prevention;
    public String treatment;
    public String severity; // "Low", "Medium", "High", "Critical"
    public String contagious; // "Yes", "No", "Unknown"
    public String mortalityRate;
    public String affectedBodyParts;
    public String seasonality;
    public String imageUrl;
    public String modelConfidence; // ML model confidence threshold
    public long dateAdded;
    public long lastUpdated;
    
    // Constructors
    public Disease() {
        this.dateAdded = System.currentTimeMillis();
        this.lastUpdated = System.currentTimeMillis();
    }
    
    public Disease(String diseaseName, String species, String symptoms, 
                  String treatment, String severity) {
        this();
        this.diseaseName = diseaseName;
        this.species = species;
        this.symptoms = symptoms;
        this.treatment = treatment;
        this.severity = severity;
    }
    
    // Getters and Setters
    public int getDiseaseId() { return diseaseId; }
    public void setDiseaseId(int diseaseId) { this.diseaseId = diseaseId; }
    
    public String getDiseaseName() { return diseaseName; }
    public void setDiseaseName(String diseaseName) { this.diseaseName = diseaseName; }
    
    public String getScientificName() { return scientificName; }
    public void setScientificName(String scientificName) { this.scientificName = scientificName; }
    
    public String getSpecies() { return species; }
    public void setSpecies(String species) { this.species = species; }
    
    public String getSymptoms() { return symptoms; }
    public void setSymptoms(String symptoms) { this.symptoms = symptoms; }
    
    public String getCauses() { return causes; }
    public void setCauses(String causes) { this.causes = causes; }
    
    public String getPrevention() { return prevention; }
    public void setPrevention(String prevention) { this.prevention = prevention; }
    
    public String getTreatment() { return treatment; }
    public void setTreatment(String treatment) { this.treatment = treatment; }
    
    public String getSeverity() { return severity; }
    public void setSeverity(String severity) { this.severity = severity; }
    
    public String getContagious() { return contagious; }
    public void setContagious(String contagious) { this.contagious = contagious; }
    
    public String getMortalityRate() { return mortalityRate; }
    public void setMortalityRate(String mortalityRate) { this.mortalityRate = mortalityRate; }
    
    public String getAffectedBodyParts() { return affectedBodyParts; }
    public void setAffectedBodyParts(String affectedBodyParts) { this.affectedBodyParts = affectedBodyParts; }
    
    public String getSeasonality() { return seasonality; }
    public void setSeasonality(String seasonality) { this.seasonality = seasonality; }
    
    public String getImageUrl() { return imageUrl; }
    public void setImageUrl(String imageUrl) { this.imageUrl = imageUrl; }
    
    public String getModelConfidence() { return modelConfidence; }
    public void setModelConfidence(String modelConfidence) { this.modelConfidence = modelConfidence; }
    
    public long getDateAdded() { return dateAdded; }
    public void setDateAdded(long dateAdded) { this.dateAdded = dateAdded; }
    
    public long getLastUpdated() { return lastUpdated; }
    public void setLastUpdated(long lastUpdated) { this.lastUpdated = lastUpdated; }
}

File: lib/features/disease/domain/entities/disease.dart
class Disease {

  const Disease({
    this.diseaseId,
    required this.diseaseName,
    required this.scientificName,
    required this.species,
    required this.symptoms,
    required this.causes,
    required this.prevention,
    required this.treatment,
    required this.severity,
    required this.contagious,
    required this.mortalityRate,
    required this.affectedBodyParts,
    required this.seasonality,
    this.imageUrl,
    required this.modelConfidence,
    required this.dateAdded,
    required this.lastUpdated,
  });

  factory Disease.create({
    required String diseaseName,
    required String species,
    required String symptoms,
    required String treatment,
    required String severity,
  }) {
    final now = DateTime.now();
    return Disease(
      diseaseName: diseaseName,
      scientificName: '',
      species: species,
      symptoms: symptoms,
      causes: '',
      prevention: '',
      treatment: treatment,
      severity: severity,
      contagious: 'Unknown',
      mortalityRate: '',
      affectedBodyParts: '',
      seasonality: '',
      modelConfidence: '0.7',
      dateAdded: now,
      lastUpdated: now,
    );
  }

  factory Disease.fromMap(Map<String, dynamic> map) {
    return Disease(
      diseaseId: map['diseaseId'],
      diseaseName: map['diseaseName'] ?? '',
      scientificName: map['scientificName'] ?? '',
      species: map['species'] ?? '',
      symptoms: map['symptoms'] ?? '',
      causes: map['causes'] ?? '',
      prevention: map['prevention'] ?? '',
      treatment: map['treatment'] ?? '',
      severity: map['severity'] ?? '',
      contagious: map['contagious'] ?? '',
      mortalityRate: map['mortalityRate'] ?? '',
      affectedBodyParts: map['affectedBodyParts'] ?? '',
      seasonality: map['seasonality'] ?? '',
      imageUrl: map['imageUrl'],
      modelConfidence: map['modelConfidence'] ?? '0.7',
      dateAdded: DateTime.fromMillisecondsSinceEpoch(map['dateAdded']),
      lastUpdated: DateTime.fromMillisecondsSinceEpoch(map['lastUpdated']),
    );
  }
  final int? diseaseId;
  final String diseaseName;
  final String scientificName;
  final String species;
  final String symptoms;
  final String causes;
  final String prevention;
  final String treatment;
  final String severity;
  final String contagious;
  final String mortalityRate;
  final String affectedBodyParts;
  final String seasonality;
  final String? imageUrl;
  final String modelConfidence;
  final DateTime dateAdded;
  final DateTime lastUpdated;

  Disease copyWith({
    int? diseaseId,
    String? diseaseName,
    String? scientificName,
    String? species,
    String? symptoms,
    String? causes,
    String? prevention,
    String? treatment,
    String? severity,
    String? contagious,
    String? mortalityRate,
    String? affectedBodyParts,
    String? seasonality,
    String? imageUrl,
    String? modelConfidence,
    DateTime? dateAdded,
    DateTime? lastUpdated,
  }) {
    return Disease(
      diseaseId: diseaseId ?? this.diseaseId,
      diseaseName: diseaseName ?? this.diseaseName,
      scientificName: scientificName ?? this.scientificName,
      species: species ?? this.species,
      symptoms: symptoms ?? this.symptoms,
      causes: causes ?? this.causes,
      prevention: prevention ?? this.prevention,
      treatment: treatment ?? this.treatment,
      severity: severity ?? this.severity,
      contagious: contagious ?? this.contagious,
      mortalityRate: mortalityRate ?? this.mortalityRate,
      affectedBodyParts: affectedBodyParts ?? this.affectedBodyParts,
      seasonality: seasonality ?? this.seasonality,
      imageUrl: imageUrl ?? this.imageUrl,
      modelConfidence: modelConfidence ?? this.modelConfidence,
      dateAdded: dateAdded ?? this.dateAdded,
      lastUpdated: lastUpdated ?? this.lastUpdated,
    );
  }

  Map<String, dynamic> toMap() {
    return {
      'diseaseId': diseaseId,
      'diseaseName': diseaseName,
      'scientificName': scientificName,
      'species': species,
      'symptoms': symptoms,
      'causes': causes,
      'prevention': prevention,
      'treatment': treatment,
      'severity': severity,
      'contagious': contagious,
      'mortalityRate': mortalityRate,
      'affectedBodyParts': affectedBodyParts,
      'seasonality': seasonality,
      'imageUrl': imageUrl,
      'modelConfidence': modelConfidence,
      'dateAdded': dateAdded.millisecondsSinceEpoch,
      'lastUpdated': lastUpdated.millisecondsSinceEpoch,
    };
  }

  @override
  String toString() {
    return 'Disease(diseaseId: $diseaseId, diseaseName: $diseaseName, scientificName: $scientificName, species: $species, severity: $severity)';
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Disease &&
        other.diseaseId == diseaseId &&
        other.diseaseName == diseaseName &&
        other.scientificName == scientificName &&
        other.species == species;
  }

  @override
  int get hashCode {
    return diseaseId.hashCode ^
        diseaseName.hashCode ^
        scientificName.hashCode ^
        species.hashCode;
  }
}
